lec04-dplyr.Rmd

```diff
@@ -1,4 +1,724 @@
+    geom_point(alpha = 0.2)
+```
+
+Great! Together with the genus plot, this definitely seem to explain most of the
+variance we see in the hindfoot length and weight measurements. It is still a
+bit messy as it appears like we have around 5 clusters, but there are 21 species
+in the legend.
+
+```{r}
+surveys %>% 
+    filter(!is.na(hindfoot_length) & !is.na(weight)) %>% 
+    group_by(species) %>% 
+    tally() %>% 
+    arrange(desc(n))
+```
+
+There is a big drop from 838 to 159, let's include only those with more than 800 observations.
+
+```{r}
+abundant_species <- surveys %>% 
+    filter(!is.na(hindfoot_length) & !is.na(weight)) %>% 
+    group_by(species) %>% 
+    tally() %>% 
+    arrange(desc(n)) %>% 
+    filter(n > 800) %>% # 
+    select(species)
+
+abundant_species
+```
+
+```{r}
+surveys_abun_species <- surveys %>% 
+    filter(!is.na(hindfoot_length) &
+        !is.na(weight) &
+        species %in% abundant_species$species)
+
+surveys_abun_species
+```
+
+Still has almost 25k observations, so only 10k was removed.
+
+```{r}
+ggplot(surveys_abun_species, aes(x = weight, y = hindfoot_length, color = species)) +
+    geom_point(alpha = 0.2)
+```
+
+
+> ### Challenge
+>
+> Create a scatter plot of `hindfoot_length` over `species` with the `weight` showing in different colors.
+> Is there any problem with this plot? *Hint: think about how many observations there are*
+
+<!--
+```{r}
+ggplot(surveys_abun_species, aes(x = weight, y = species, color = hindfoot_length)) +
+    geom_point(size = 0.1, position = 'jitter')
+```
+-->
+
+
+### Split-apply-combine... plot!
+
+In this section, we will learn how to work with `**dplyr**` and `**ggplot**`
+together. Aided by the pipes (`%>%`), we can create a powerul data exploration
+workflow using these two packages.
+
+Let's calculate number of counts per year for each species. First we need
+to group the data and count records within each group:
+
+```{r}
+surveys_abun_species %>%
+    group_by(year, genus) %>%
+    tally() %>% 
+    arrange(desc(n)) # Adding arrange just to compare with histogram
+```
+
+We could assign this table to a variable, and then pass that variable to `ggplot()`.
+
+```{r}
+yearly_counts <- surveys_abun_species %>%
+    group_by(year, species) %>%
+    tally()
+
+ggplot(yearly_counts, aes(x = n)) +
+    geom_histogram()
+```
+
+Creating an intermediate variable, would be prefereable if you are preferring a time consuming calculation, because you would not want to do that operation everytime you change some plot aesthetic.
+
+If it is not a time consuming approach and you would like the flexibility of
+changing the data summary and the plotting options in the same code chunk, you
+can pipe the output of your split-apply-combine operation to the plotting
+command:
+
+```{r}
+surveys_abun_species %>%
+    group_by(year, species) %>%
+    tally() %>% 
+    ggplot(aes(x = n)) +
+        geom_histogram()
```

Comment:
geom_ should be at the same level as the other commands (same indentation)

Reply:
What does `Ctrl-Shift-A` do to the code?

Reply:
@lwjohnst86 `Ctrl-Shift-a` is on your side, but I disagree with the both of you =)

My reasoning is that if the `ggplot` call would not be in a chain, `geom_histogram` would be indented:

```R
ggplot(aes(x = n)) +
    geom_histogram()
```

That indentation conveys the relationship between `geom_histogram` and `ggplot`, and this relationships remains the same when used in a chain.

To me, this is a similar situation to what is described in [4.4 Long lines](http://style.tidyverse.org/pipes.html#long-lines-2) of the tidyverse style guide.

In the last example of [7.7 ggplot2 calls](http://r4ds.had.co.nz/exploratory-data-analysis.html#ggplot2-calls) in R for Data Science, they use the same syntax as I suggested:

```{R}
diamonds %>%
  count(cut, clarity) %>%
  ggplot(aes(clarity, cut, fill = n)) +
    geom_tile()
```

`Ctrl-Shift-a` also changes the chunk below

```{r}
surveys_abun_species %>%
    summarize(
        mean_weight = mean(weight),
        min_weight = min(weight))
```

into this


```{r}
surveys_abun_species %>%
    summarize(mean_weight = mean(weight),
    min_weight = min(weight))
```

which does not adhere to the tidyverse style advice for long lines linked above. Is `Ctrl-Shift-a` linter for general R syntax, but does not follow the specific tidyverse recommendations?